### PR TITLE
INNO: Generate analytics whitelist at build time

### PR DESCRIPTION
**Pull request by tarka**
There is already drift from our actual analytics and the whitelist; it's better to generate the whitelist from the known analytics sources.

### CODE REVIEW

File: jira-plugin/src/main/kotlin/com/atlassian/migration/datacenter/build/GenWhitelist.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * Small util to generate an analytics whitelist from the defined events.
+ * Intended to be called during the build phase; see `pom.xml` and `atlassian-plugin.xml`.
+ */
+fun main(args: Array<String>) {
+    val pkg = args[0]
+    val target = args[1]
+
+    val eventAnno = "com.atlassian.analytics.api.annotations.EventName"
```
Comment (vicav) [inline jira-plugin/src/main/kotlin/com/atlassian/migration/datacenter/build/GenWhitelist.kt:31]: You should be able to just use spring for this: https://stackoverflow.com/questions/259140/scanning-java-annotations-at-runtime

File: jira-plugin/pom.xml [hunk 2]
```diff
@@ -83,6 +83,19 @@
             <artifactId>atlassian-plugins-osgi-javaconfig</artifactId>
         </dependency>
 
+        <dependency>
```
Comment (bpar476) [inline jira-plugin/pom.xml:86]: Do these dependencies get compiled into the final JAR?

File: jira-plugin/src/main/kotlin/com/atlassian/migration/datacenter/build/GenWhitelist.kt (hunk 1 above)
Comment (tarka) [reply jira-plugin/src/main/kotlin/com/atlassian/migration/datacenter/build/GenWhitelist.kt:31]: I've used Spring Scanner in the past, but the when using it with a standalone executable you rapidly get into classpath hell, as a `provided` dependency is visible to it. ClassGraph is smaller and simpler to work with so made sense in this case.

File: jira-plugin/pom.xml (hunk 2 above)
Comment (tarka) [reply jira-plugin/pom.xml:86]: Not anymore, I just fixed that.

File: jira-plugin/src/main/kotlin/com/atlassian/migration/datacenter/build/GenWhitelist.kt (hunk 1 above)
Comment (tarka) [reply jira-plugin/src/main/kotlin/com/atlassian/migration/datacenter/build/GenWhitelist.kt:31]: Actually, that's not correct, it is possible to use a provided dependency. However ClassGraph is simpler to work with, so I'm happy to leave it in as it doesn't cost anything.

Comment (vicav) [reply jira-plugin/src/main/kotlin/com/atlassian/migration/datacenter/build/GenWhitelist.kt:31]: Yep - now that the deps are provided and not compile scope agreed.

Comment (tarka) [reply jira-plugin/src/main/kotlin/com/atlassian/migration/datacenter/build/GenWhitelist.kt]: So making the dependencies `provided` confuses OSGI and the whole thing falls over. I had to refactor them all out into a separate `buld-utils` module, which probably makes more sense anyway.

Comment (tarka) [reply jira-plugin/src/main/kotlin/com/atlassian/migration/datacenter/build/GenWhitelist.kt]: If this is of use to ATST we can split it out into its own repo.

File: build-utils/pom.xml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            <artifactId>classgraph</artifactId>
+            <version>${classgraph.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
+            <version>${gson.version}</version>
+        </dependency>
+    </dependencies>
+
```
Comment (vicav) [inline build-utils/pom.xml:48]: Should move the gson.version, classgraph.version properties to this pom too

File: jira-plugin/pom.xml [hunk 4]
```diff
@@ -88,10 +88,45 @@
             <artifactId>junit-jupiter-api</artifactId>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>com.atlassian.migration.datacenter</groupId>
+            <artifactId>build-utils</artifactId>
+            <version>${project.version}</version>
+            <scope>provided</scope>
```
Comment (theghostwhoforks) [inline jira-plugin/pom.xml:96]: Is this set to `provided` as we don't want to compile this into the generated jar/obr?

Comment (tarka) [reply jira-plugin/pom.xml:96]: Correct; the module and dependencies are downloaded and available at bulid time, but not packaged.

**Pull request merged**